File: 3Sum/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        st = set()
        
        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j, k = i + 1, n - 1
            
            while j < k:
                total = nums[i] + nums[j] + nums[k]
                if total == 0:
                    st.add((nums[i], nums[j], nums[k]))
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
                elif total > 0:
                    k -= 1
                else:
                    j += 1

        return list(map(list, st))
        

File: 3Sum/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        int n = nums.size();
        set<vector<int>> st;
        int i =0;
        while(i<n) {
            int j = 0,k = n-1;
            while(j<k) {
                if(j == i) {
                    j++;
                    continue;
                }
                else if(i == k) {
                    k--;
                    continue;
                }
                int sum = nums[i]+nums[j]+nums[k];
                if(sum == 0) {
                    vector<int> temp= {nums[i],nums[j],nums[k]};
                    sort(temp.begin(),temp.end());
                    st.insert(temp);
                    j++;
                    k--;
                }
                else if(sum > 0) k--;
                else j++;
            }
            int temp = nums[i];
            while(i<n && nums[i] == temp) i++;
        }
        vector<vector<int>> ans;
        for(auto i: st) {
            ans.push_back(i);
        }
        return ans;
    }
};